[N7 Account Settings] Create personal details pages

Beamanator opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->
cc @arosiclair - We can use this PR to test out your PR to create the commands. This is almost ready for testing, I have a bit more testing and cleanup to do first

### Details
<!-- Explanation of the change or anything fishy that is going on -->
~Holding on Web-E PR that creates the necessary "update" commands: https://github.com/Expensify/Web-Expensify/pull/36075~
- ~Update: Holding on https://github.com/Expensify/Web-Expensify/pull/36284~

### Fixed Issues
$ https://github.com/Expensify/App/issues/11578
$ https://github.com/Expensify/App/issues/11579
$ https://github.com/Expensify/App/issues/11580
$ https://github.com/Expensify/App/issues/11581

### Tests

**Note when testing in dev only**: Comment out [this line](https://github.com/Expensify/Web-Expensify/blob/69fa400404c62927c9af5c15410f4c7ba6610c09/lib/PersonalDetailsAPI.php#L314) in web-e so local dev testing works!
- Actually I think the problem here is related to [this thread](https://expensify.slack.com/archives/C02HWMSMZEC/p1674658845553099)

#### Profile page
1. Navigate to Settings -> Profile page
2. Verify all of the menu item heights look the same and everything looks good, like this screenshot:
    - <img width="816" alt="Screenshot 2023-01-20 at 2 27 26 PM" src="https://user-images.githubusercontent.com/3885503/213694806-358ffd60-a1a4-4b3c-a7b2-3e2864894952.png">
4. Verify clicking on Personal Details button navigates to personal details initial page

#### Personal Details initial page (part 1)
1. Navigate to Settings -> Profile -> Personal Details
2. Log in to an account with no legal name, date of birth, or address set verify this page looks empty like this:

<img width="369" alt="Screenshot 2023-01-17 at 3 07 33 PM" src="https://user-images.githubusercontent.com/3885503/212906921-c444ee53-360a-4705-be27-3b3fdc18cbcb.png">

#### Legal name page
1. Navigate to Settings -> Profile -> Personal Details -> Legal name
2. Verify you cannot submit the form with either of the legal name fields empty
3. Verify you see an error below fields if either field has > 150 characters
    - <img width="638" alt="Screenshot 2023-01-20 at 2 29 55 PM" src="https://user-images.githubusercontent.com/3885503/213695339-4168cc5f-9db0-4e5a-b796-1ab4b1bb6121.png">
4. Verify you see an error below fields if either field has characters `,` or `;`
    - <img width="638" alt="Screenshot 2023-01-20 at 2 30 28 PM" src="https://user-images.githubusercontent.com/3885503/213695395-27ca166f-c2a7-48e9-86da-7e0df84c8104.png">
5. Verify when you click "Save" you are redirected back to the Personal Details Initial page, and that you now see the updated legal name (first + last with a space inbetween)
    - Verify you see the update on a second logged in device too, unless offline

#### Date of birth page
1. Navigate to Settings -> Profile -> Personal Details -> Date of birth
2. Fill out a date less than 5 years ago. Click save, verify you see an error that makes sense
    - <img width="443" alt="Screenshot 2023-01-20 at 2 42 59 PM" src="https://user-images.githubusercontent.com/3885503/213697629-3b813df3-ad22-4ace-bd6a-78bbfb4d0df4.png">
3. Fill out a date more than 150 years ago. Click save, verify you see an error that makes sense
    - <img width="443" alt="Screenshot 2023-01-20 at 2 45 51 PM" src="https://user-images.githubusercontent.com/3885503/213698116-984d0dd7-4e40-476b-993c-9f4afc9d49bb.png">
5. Fill out a date that is invalid (like November 31 2000). Click save, verify you see an error that makes sense
    - <img width="443" alt="Screenshot 2023-01-20 at 2 46 20 PM" src="https://user-images.githubusercontent.com/3885503/213698207-e83646d6-668d-4a7b-accc-4a037d2adc2b.png">
6. Fill out a date that is valid (and between 5 and 150 years ago), make sure you see no error
7. Click "Save", verify you are redirected back to the Personal Details Initial page, and that you now see the updated date of birth

#### Address page
1. Navigate to Settings -> Profile -> Personal Details -> Home address
2. Verify you cannot submit the form with **any** of the address fields empty except the "Address line 2" field (an error shows up on all fields that are empty, except address line 2)
    - <img width="443" alt="Screenshot 2023-01-20 at 2 47 28 PM" src="https://user-images.githubusercontent.com/3885503/213698410-e7a2b58a-3ade-4db4-a1be-4746438b0f17.png">
3. Verify you cannot enter more than 50 characters on any field (except the first address line & you only enter up to 10 in the zip code field)
4. Verify if you have "USA" selected in the country picker, the "State" picker is shown. But if you switch to any other country, the "State / Province" Text input field now shows
    - https://user-images.githubusercontent.com/3885503/213698697-dfc0a665-bdf0-4cce-be7d-1f389b5d1c98.mov
5. Fill out the form with valid data, click "Save", verify you are redirected back to the Personal Details Initial page, and that you now see the updated address on that page - all fields should be separated by commas
6. Go to the home address page, start typing in the street field, click on a suggested address
7. Verify the fields in the form get auto-populated

#### Personal Details initial page (part 2)
1. Navigate to Settings -> Profile -> Personal Details
8. Log in to an account with legal name, date of birth, or address set. Verify this page looks like this:

<img width="386" alt="Screenshot 2023-01-20 at 2 49 55 PM" src="https://user-images.githubusercontent.com/3885503/213698860-9980b7d6-9ad0-4559-b3b7-1644a84b69a0.png">

- [ ] Verify that no errors appear in the JS console

### Offline tests
You should be able to do all of the above offline, except looking at updates on a 2nd device

### QA Steps
Same as above

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/3885503/215795934-4b7c0a0f-e250-4245-becf-934bdc0bc9fa.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/3885503/216363350-b209aa9e-8b61-4930-9da7-9173677899ef.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/3885503/216344040-f826db2d-4391-4171-8187-e1580f75f149.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/3885503/216360468-2d06c52c-8986-4961-853a-b4d88dd03b85.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/3885503/216342996-f9295e5f-4360-44b1-92a1-3fd20eda71a1.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/3885503/216362504-9d1dab7f-2f92-42d4-a86a-b499b25b329f.mov

</details>

Beamanator committed:
Set up routes for new personal details pages

Beamanator committed:
New initial personal details page

Beamanator committed:
Add legal name page

Beamanator committed:
Fix personal details page init linking

Beamanator committed:
Add date of birth page

Beamanator committed:
New Home Address page

Beamanator committed:
Country picker initial commit

Beamanator committed:
Make variable names more clear

Beamanator committed:
Add button to navigate to personal details page

Beamanator committed:
Add all legal first name errors

Beamanator committed:
Make variable names a bit clearer

Beamanator committed:
Add necessary default & trim

Beamanator committed:
Modify validation fn to be dynamic

Beamanator committed:
Start creating dob validation fn

Beamanator committed:
Revert since we'll make a new function

Beamanator committed:
Make comment clearer

Beamanator committed:
Remove unused imports

Beamanator committed:
Add better date errors

Beamanator committed:
Move everything to new private_personalDetails key

Beamanator committed:
Finish up Address field validation

Beamanator committed:
Remove unused imports

Beamanator committed:
Clean up required field validation

Beamanator committed:
Merge branch 'main' of github.com:Expensify/App into beaman-createPersonalDetailsPages

Beamanator committed:
Fix api endpoints, body data, and cleanup

Beamanator committed:
Clean up a lot

Beamanator committed:
Merge branch 'main' of github.com:Expensify/App into georgia-new-avatars

Beamanator commented:
Few more things to figure out before this should be 100% ready:
1. Pusher updates when updating DOB & Address
2. Country picker

Beamanator committed:
Allow google places to populate address field

Beamanator committed:
Update translation key

Beamanator committed:
Add list of all countries for country picker

Beamanator committed:
Address lint errors

Beamanator commented on src/languages/en.js:
```diff
@@ -552,12 +559,28 @@ export default {
+        dateOfBirth: 'Date of birth',
+        homeAddress: 'Home address',
+        error: {
+            hasInvalidCharacter: ({characterName}) => `Please remove the ${characterName} from the field above.`,
+            legalFirstNameEmpty: 'Please enter a valid legal first name',
+            legalLastNameEmpty: 'Please enter a valid legal last name',
+            dateInvalid: 'Please enter a valid date.',
+            dateTooRecent: ({numYears}) => `Date needs to be less than ${numYears} years ago.`,
+            dateTooOld: ({numYears}) => `Date cannot be more than ${numYears} years ago.`,
```
@shawnborton sorry for tagging you all over recently, would you mind looking at these error messages to see if they make sense for our upcoming Legal name page & Date of birth page?

Maybe I can get rid of the trailing `.` since that's inconsistent here, otherwise are they decent?

Beamanator commented on src/CONST.js:
```diff
@@ -881,6 +882,258 @@ const CONST = {
     TFA_CODE_LENGTH: 6,
     CHAT_ATTACHMENT_TOKEN_KEY: 'X-Chat-Attachment-Token',
+
+    ALL_COUNTRIES: [
+        'Afghanistan',
+        'Aland Islands',
+        'Albania',
+        'Algeria',
+        'American Samoa',
```
Might need to get spanish translation of all these 😬

MonilBhavsar replied:
👍  I guess, we'll store only english value in storage.
Also, I would recommend adding a test to check that this list has equal number of entries in both english and Spanish translations and compare values at random indexes.

Beamanator replied:
Hmm very true, I like those ideas - but also that got me thinking, we only show States in the StatePicker in English, right? Maybe a follow-up could be translating both of those?

MonilBhavsar replied:
Ideally, these countries should be in spanish if selected. Do we display states, countries right now? I don't think so

Beamanator replied:
As we discussed 1:1 - We don't display countries yet, states so far we only display the state two letter code, which probably doesn't have a translation 😅

Beamanator replied:
Made a follow-up for this since I don't want to hold this PR back much longer: https://github.com/Expensify/App/issues/14549

Beamanator commented:
@eVoloshchak FYI this PR is basically ready for lots of testing, I just have a bit of translation work to do before it's ready to merge - so if you have time to test it soon that would be great 🙏

eVoloshchak commented:
> @eVoloshchak FYI this PR is basically ready for lots of testing, I just have a bit of translation work to do before it's ready to merge - so if you have time to test it soon that would be great 🙏

Nice! I'll start testing later today

arosiclair commented on src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.js:
```diff
@@ -0,0 +1,116 @@
+
+    /** User's private personal details */
+    privatePersonalDetails: PropTypes.shape({
+        dateOfBirth: PropTypes.string,
```
I think this'll get set as `dob` here in Onyx as well

arosiclair reviewed (commented):
Looks pretty good so far!

Beamanator committed:
Merge branch 'main' of github.com:Expensify/App into beaman-createPersonalDetailsPages

MonilBhavsar commented on src/ROUTES.js:
```diff
@@ -42,6 +42,10 @@ export default {
     SETTINGS_PAYMENTS_TRANSFER_BALANCE: 'settings/payments/transfer-balance',
     SETTINGS_PAYMENTS_CHOOSE_TRANSFER_ACCOUNT: 'settings/payments/choose-transfer-account',
+    SETTINGS_PERSONAL_DETAILS: 'settings/profile/personal-details',
+    SETTINGS_PERSONAL_DETAILS_LEGAL_NAME: 'settings/profile/personal-details/legal-name',
+    SETTINGS_PERSONAL_DETAILS_DATE_OF_BIRTH: 'settings/profile/personal-details/date-of-birth',
+    SETTINGS_PERSONAL_DETAILS_ADDRESS: 'settings/profile/personal-details/address',
```
```suggestion
    SETTINGS_PERSONAL_DETAILS_LEGAL_NAME: `${SETTINGS_PERSONAL_DETAILS}/legal-name`,
    SETTINGS_PERSONAL_DETAILS_DATE_OF_BIRTH: `${SETTINGS_PERSONAL_DETAILS}/date-of-birth`,
    SETTINGS_PERSONAL_DETAILS_ADDRESS: `${SETTINGS_PERSONAL_DETAILS}/address`,
```

Beamanator replied:
hadn't seen us doing this before! i like it!

MonilBhavsar replied:
We've been doing this for IOU and other routes 😄

MonilBhavsar commented on src/components/CountryPicker.js:
```diff
@@ -0,0 +1,66 @@
+    /** The value that needs to be selected */
+    value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
+
+    /** The ID used to uniquely identify the input in a Form */
```
```suggestion
    /** The ID used to uniquely identify the input in a form */
```

MonilBhavsar reviewed (commented):
Some comments after a brief first pass

Beamanator committed:
Language updates

Beamanator commented on src/libs/actions/PersonalDetails.js:
```diff
@@ -173,6 +173,74 @@ function updateDisplayName(firstName, lastName) {
+            onyxMethod: CONST.ONYX.METHOD.MERGE,
+            key: ONYXKEYS.PRIVATE_PERSONAL_DETAILS,
+            value: {
+                dateOfBirth,
```
I actually fixed this by sending `dob: dateOfBirth` above, but yeah I agree this looks bad and inconsistent so will fix 👍

Beamanator committed:
Make dob vars match API

Beamanator committed:
Merge branch 'main' of github.com:Expensify/App into beaman-createPersonalDetailsPages

Beamanator commented on src/pages/settings/Profile/PersonalDetails/AddressPage.js:
```diff
@@ -0,0 +1,223 @@
+        this.updateAddress = this.updateAddress.bind(this);
+        this.onCountryUpdate = this.onCountryUpdate.bind(this);
+
+        const currentCountry = lodashGet(props.privatePersonalDetails, 'address.country') || '';
```
See I was doing that, but somehow during testing I was getting `null` here 'n there, and sadly `lodashGet`'s default param doesn't get applied with the value is `null`, only when it's `undefined`

MonilBhavsar replied:
I see 👍

Beamanator committed:
Make personal details routes reuse prefix

Beamanator committed:
Move United States name to const

Beamanator committed:
Comment updates

Co-authored-by: Monil Bhavsar <monil@expensify.com>

Beamanator commented on src/libs/ValidationUtils.js:
```diff
@@ -206,6 +206,30 @@ function meetsAgeRequirements(date) {
+    if (!testDate.isValid()) {
+        return Localize.translateLocal('privatePersonalDetails.error.dateInvalid');
+    }
+    if (testDate.isBetween(longAgoDate, recentDate)) {
+        return '';
+    }
```
This is the only "valid" date case - I wonder if it would be clearer if this was at the bottom of this function?

Beamanator committed:
Minor logic fix

Beamanator committed:
Merge branch 'beaman-createPersonalDetailsPages' of github.com:Expensify/App into beaman-createPersonalDetailsPages

Beamanator commented on src/languages/es.js:
```diff
@@ -552,12 +559,27 @@ export default {
+    privatePersonalDetails: {
+        personalDetails: 'Datos personales',
+        privateDataMessage: 'Estos detalles se utilizan para viajes y pagos. Nunca se mostrarán en su perfil público.',
+        legalName: 'Nombre completo',
+        legalFirstName: 'Nombre',
+        legalLastName: 'Apellidos',
+        homeAddress: 'Domicilio',
+        error: {
+            hasInvalidCharacter: '',
+            legalFirstNameEmpty: '',
+            legalLastNameEmpty: '',
```
Schmuck... I'm going to deal with cleaning up some of these error messages tomorrow

eVoloshchak commented:
@Beamanator, I've found a couple of issues

1. Not sure if it's a bug with my environment, but when navigating to `PersonalDetailsInitialPage`, app crashes with the following error:
<img width="676" alt="image" src="https://user-images.githubusercontent.com/9059945/214555428-330a5c7e-be8d-4e0d-a5b6-7c822c371bbe.png">

> Invalid attempt to destructure non-iterable instance

It doesn't matter if you have a legal name or not. 
This line is at fault

https://github.com/Expensify/App/blob/3de46a3a807ed0e4eefacc1d975067121b3bd7b2/src/pages/settings/Profile/PersonalDetails/PersonalDetailsInitialPage.js#L78

Using the same approach you used in `AddressPage` has resolved the issue

https://github.com/Expensify/App/blob/3de46a3a807ed0e4eefacc1d975067121b3bd7b2/src/pages/settings/Profile/PersonalDetails/AddressPage.js#L125

2. Legal name isn't updated until you close the Personal Details page and then open it again. Not sure if it should be polled and updated automatically or updated only when leaving/coming back to  Personal Details page

https://user-images.githubusercontent.com/9059945/214559207-e83e5da6-b546-4690-9740-159613479681.mov

3. Initially, inputs on Legal Name page don't validate for errors as you type, only when the input is blurred. This is until the input is blurred for the first time, after that the validation occurs for every symbol as you type. iOS and Android also have this problem

 

https://user-images.githubusercontent.com/9059945/214560114-1667cf80-e475-4c5a-92c3-48e4b874053c.mov

Beamanator committed:
Merge branch 'main' of github.com:Expensify/App into beaman-createPersonalDetailsPages

Beamanator commented:
Thanks for the tests @eVoloshchak ! Some responses:

> 1. Not sure if it's a bug with my environment, but when navigating to PersonalDetailsInitialPage, app crashes with the following error:

Odd, I actually haven't seen that error on my side but I'll update and see if it fixes it for you 👍 

> 2. Legal name isn't updated until you close the Personal Details page and then open it again. Not sure if it should be polled and updated automatically or updated only when leaving/coming back to Personal Details page

This is interesting, to get this to work in my dev environment I have to comment out some expensify card user data updating... I wonder if that's throwing for you... I'll look in the logs soon

> 3. Initially, inputs on Legal Name page don't validate for errors as you type, only when the input is blurred. This is until the input is blurred for the first time, after that the validation occurs for every symbol as you type. iOS and Android also have this problem

Yeah, this is actually expected, it's the way the forms were designed - I brought this up [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1669633176439899?thread_ts=1659440463.665489&cid=C01GTK53T8Q) but the point is that there are many fields that we don't want to validate _on each keystroke_ (email validation, for example) so I believe we're planning to keep this as is

Beamanator commented on src/languages/es.js:
```diff
@@ -552,12 +559,27 @@ export default {
+        legalLastName: 'Apellidos',
+        homeAddress: 'Domicilio',
+        error: {
+            hasInvalidCharacter: '',
```
Working on this one here: https://expensify.slack.com/archives/C21FRDWCV/p1674231085985749

Beamanator commented on src/languages/es.js:
```diff
@@ -552,12 +559,27 @@ export default {
+            hasInvalidCharacter: '',
+            legalFirstNameEmpty: '',
+            legalLastNameEmpty: '',
+            dateInvalid: 'Ingresa una fecha válida',
+            dateTooRecent: ({numYears}) => ``,
+            dateTooOld: ({numYears}) => ``,
```
Working on these 3 soon...

Beamanator replied:
https://expensify.slack.com/archives/C21FRDWCV/p1674661149364599

Beamanator committed:
Lint fixes

Beamanator committed:
A few error updates

Beamanator committed:
Combine common invalid date error message

Beamanator commented:
@eVoloshchak - yeah about your second point, I started an internal discussion ([here](https://expensify.slack.com/archives/C02HWMSMZEC/p1674658845553099), which you don't have access to sorry) - for now you probably can't do any of the tests 😅

Beamanator committed:
Final bit of error message cleanup

Beamanator commented:
@MonilBhavsar @arosiclair it looks like @eVoloshchak is going to have trouble testing until I implement the plan detailed [here](https://expensify.slack.com/archives/C02HWMSMZEC/p1674659021452629) (hopefully tomorrow, but won't be in staging till early next week).

I also finished up the error messages & made this follow-up (https://github.com/Expensify/App/issues/14549) for country picker translations. So if y'all could review again by EOW that would be super helpful 👍 🙏

Beamanator committed:
Address lint errors

Beamanator committed:
Please don't make me strangle anyone

arosiclair reviewed (dismissed):
Looking pretty good and tests well on web!

Beamanator commented:
Putting on hold for https://github.com/Expensify/Web-Expensify/pull/36284, where we'll handle the pusher bug mentioned by @eVoloshchak (updating private personal details in a new account used to throw errors because we tried updating Marqeta, but user didn't have all required private personal details)

Beamanator commented:
@eVoloshchak Actually I think you can continue tests, just assume on a 2nd device you won't see immediate updates 😅

Beamanator committed:
Merge branch 'main' of github.com:Expensify/App into beaman-createPersonalDetailsPages

eVoloshchak commented:
> @eVoloshchak Actually I think you can continue tests, just assume on a 2nd device you won't see immediate updates 😅 (until https://github.com/Expensify/Web-Expensify/pull/36284 is merged)

Awesome, will continue testing later today

Beamanator commented:
One more "issue" I'm considering making a follow-up: When you save a legal name (for example), I believe we're retriggering calling `OpenPersonalDetailsInitialPage` which results in:
1. Optimistic onyx data is stored from updating legal name
2. Results from `OpenPersonalDetailsInitialPage` return and get displayed, which is old data
3. Pusher message from updating legal name arrives, with new data

Beamanator committed:
Improve date validation errors

Beamanator commented:
No longer holding on Web-E PR! 👍

eVoloshchak commented:
There are a couple of crashes due to localization 
1. There is no `common.error.comma` and `common.error.semicolon`, so it crashes on legal name screen is you enter comma or semicolon
2. There is no `privatePersonalDetails.error.dateInvalid`, so there's a crash when you try to delete a date

https://user-images.githubusercontent.com/9059945/215760091-2689f0e0-43b0-4d2b-8d4d-3b6ef8e66f46.mov

3. One real bug I've found is that if the address line exceeds 150 characters, the form will still be submitted, but the address won't be updated


https://user-images.githubusercontent.com/9059945/215761366-257e028c-810c-4ab2-bcce-21eda3868368.mov

4. Not sure if that's the intended behavior, but date picker isn't cleared when you leave the page, while legal name and address pages are cleared


https://user-images.githubusercontent.com/9059945/215762188-1e051abd-e60b-43fe-adef-bf65a1eedcdb.mov


Other than that, I've tested these flows on all platforms and it looks and works well

eVoloshchak commented:
I also have a suggestion for the address form, but not sure if that's going to be hard to implement: we could automatically select a country in the country picker when user selects the auto-suggested address

https://user-images.githubusercontent.com/9059945/215763727-8a1c4245-e886-4b7d-9b75-43545da941f5.mov

Beamanator committed:
Fix a few error messages

Beamanator committed:
Add ability to cap address search at 50 characters

Beamanator committed:
Merge branch 'main' of github.com:Expensify/App into beaman-createPersonalDetailsPages

Beamanator committed:
Use standard moment date format for all platforms

Beamanator commented:
@eVoloshchak did you experience any weird behavior in the date of birth page on iOS? For some reason I'm getting errors and I think it's because `values.dob` is a `Date` object instead of a string??

eVoloshchak commented:
> @eVoloshchak did you experience any weird behavior in the date of birth page on iOS? For some reason I'm getting errors and I think it's because `values.dob` is a `Date` object instead of a string??

Hmm, actually I'm getting this error on both iOS and Android when pressing `Save` 
I must have missed this, but I remember datepicker working on native at some point

![simulator_screenshot_A31E3005-391C-45F1-982C-D3B1689DACA5](https://user-images.githubusercontent.com/9059945/215816932-a73bbd9c-f723-44c7-b622-75a926c18a10.png)

Beamanator commented:
@eVoloshchak yeah that's what i'm seeing too - but not always I think 🤔  I don't think this is related to my changes in this PR though, since I didn't adjust the DatePicker & I didn't use it in a weird way 😅

eVoloshchak commented:
> Do you mind testing this on a DatePicker that's in `main`?

Pulled latest `main`, DatePicker on Personal Information screen works fine on Android 
I can't check iOS since Passwordless is too passwordless and I'm not receiving sign-in links

eVoloshchak commented:
> Sorry y'all I'll finish the checklist & screenshots tomorrow 🙏 Then are we good to go with tests & a few follow-ups @eVoloshchak ?

Yes!

Beamanator committed:
Merge branch 'main' of github.com:Expensify/App into beaman-createPersonalDetailsPages

Beamanator commented:
The iOS date issue i mentioned before is gone 😬 Dunno how it happened 🤷

Beamanator commented:
Screenshots & vids all uploaded from me, @eVoloshchak @arosiclair mind giving it a final review today? 🙏

eVoloshchak commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

eVoloshchak commented:
@Beamanator, tested again and uploaded the screenshots
I'm still getting the date error on iOS and Android, so can't fully test those. I've verified that datepicker in personal information screen

Beamanator commented:
@eVoloshchak 

> I'm still getting the date error on iOS and Android, so can't fully test those. I've verified that datepicker in personal information screen

Ugg that's annoying, but I think we should make it a follow-up since I don't think it has anything to do with my changes, it must be a generic date picker bug on native (I didn't see it on Android by the way, only iOS)

> Sorry for noticing this just now, is ZIP code input supposed to not be validated? I can enter anything into it

Good question, for now I think we want to keep it as it is, since some countries have letters and / or numbers in the zip code. Example:

> 6 Christchurch Terrace, London SW3 4AJ, United Kingdom

https://goo.gl/maps/XcNWtuuJfVCeL4zW7

eVoloshchak commented:
@Beamanator, makes sense
In that case, I'll re-test and upload screenshots for native in an hour, everything else looks good

Beamanator commented:
Groovy, thanks @eVoloshchak 👍 👍

MonilBhavsar commented on src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.js:
```diff
@@ -0,0 +1,118 @@
+        return (
+            <ScreenWrapper includeSafeAreaPaddingBottom={false}>
+                <HeaderWithCloseButton
+                    shouldShowBackButton
```
Missing `title` here

Beamanator replied:
yeahhhh this is actually the design 😅  https://docs.google.com/document/d/17akqrbC-w8FqEoDLvcLmrX6iw4uaVrT_xlwAl-IhEzE/edit#bookmark=kix.yb99cel3cv0k

MonilBhavsar replied:
I see. A design bug then 😅 Every screen has a title, so I think it was missed in this case

Beamanator replied:
Sadly it's not so simple 😅 This was decided because the page only has one input, so people argued that the page title would be redundant 🤷  I can get people involved who made the decision

MonilBhavsar replied:
Not a blocking change, We can move forward, discuss and address later. But I feel like this doesn't match with other fields.

> This was decided because the page only has one input, so people argued that the page title would be redundant

Oh, But the input label says "Date" only, so I think it might be good to have a title that explains which date it is and also align with other pages.

MonilBhavsar reviewed (commented):
Looks good overall!

eVoloshchak commented:
Found a small issue, `Personal Details` page isn't scrollable 

https://user-images.githubusercontent.com/9059945/216588519-6ac9019b-96c2-4c00-93a6-e65708897818.mov

Beamanator committed:
Merge branch 'main' of github.com:Expensify/App into beaman-createPersonalDetailsPages

Beamanator committed:
Add scrollview to page

Beamanator commented:
@eVoloshchak good catch, added the scrollview 👍

Beamanator committed:
(you saw nothing)

MonilBhavsar commented:
I saw a small validation issue with Address page.
Initially if I don't do anything with country selector, it accepts it and probably stores wrong value.


https://user-images.githubusercontent.com/32012005/216609538-558656d4-5c48-4bf0-b9be-e94591691f96.mov

MonilBhavsar commented:
Sorry, another issue with country selector not updating with selected/latest value


https://user-images.githubusercontent.com/32012005/216610810-d19b9241-4209-47b1-b0f3-6e2f89c5aa20.mov

Beamanator commented:
@MonilBhavsar 

> I saw a small validation issue with Address page.
Initially if I don't do anything with country selector, it accepts it and probably stores wrong value.

Hmm I have only reproduced when I accidentally had something other than a valid country stored in Onyx - is that how you reproduced that one?

> Sorry, another issue with country selector not updating with selected/latest value

I thinkkkkk this is due to this: https://github.com/Expensify/App/issues/14784

MonilBhavsar commented:
> Hmm I have only reproduced when I accidentally had something other than a valid country stored in Onyx - is that how you reproduced that one?

I am able to reproduce it every time. The country also doesn't look correct, any idea how false country shows up? Also, @eVoloshchak Are you able to reproduce this issue?

> I thinkkkkk this is due to this: https://github.com/Expensify/App/issues/14784

Looks like it. Thanks! 👍

Beamanator commented:
@MonilBhavsar - I think the reason the country picker validation sometimes doesn't work is b/c I used to have "AF" stored in country, I think from old test. When I updated to a valid country, I saw everything work properly, so I wonder if that's the same for you

Also in your test I can see that the country started as "AF" - I think that was a starting test value before I added the full list: 

<img width="315" alt="Screenshot 2023-02-03 at 4 12 21 PM" src="https://user-images.githubusercontent.com/3885503/216624854-0df22e4c-996b-40de-b396-1b7239004c9f.png">

I've been trying to break it but can't so far, so I wonder if you can break **after** you successfully set a real country in the picker?

eVoloshchak commented:
> @eVoloshchak Are you able to reproduce this issue?

Yes, was able to reproduce on a new account

https://user-images.githubusercontent.com/9059945/216632960-e1894510-5ce2-4554-9c52-27ec4a093183.mov

Beamanator commented:
Ahhhhhhhh @eVoloshchak your video helped 👍  The reason this happens is b/c google places DOES populate the country, but the country is only two letters (in your example, `US`) and the country picker has different values, so it shows up as `-` but technically the country is populated - I'll fix this somehow 🙃

Beamanator committed:
Fix country auto-population

Beamanator commented:
Fixed - please hopefully a final review @MonilBhavsar @eVoloshchak 🙏 🙏 🙏

eVoloshchak commented:
Re-tested and added recording for Android

Encountered an issue on iOS: I logged out, signed back with a new account, the address field was empty, but when I navigated to Home address screen it was populated by the info from the previous account. 
 

https://user-images.githubusercontent.com/9059945/216851236-9afb57f4-8894-4158-bcea-add4842c5923.mov


Also if you edit home address and then come back to edit it again, the changes aren't saved on home address page
 

https://user-images.githubusercontent.com/9059945/216851273-1f870b27-cbe9-40d3-9e1e-5934d141d6ec.mov

This is also happening on a physical device

Beamanator commented:
@eVoloshchak I'm 99% sure those issues you mentioned are `Form` bugs, unrelated to my PR - but I'll try to dive in to them today 👍

Beamanator committed:
Merge branch 'main' of github.com:Expensify/App into beaman-createPersonalDetailsPages

Beamanator commented:
> Encountered an issue on iOS: I logged out, signed back with a new account, the address field was empty, but when I navigated to Home address screen it was populated by the info from the previous account.

> Also if you edit home address and then come back to edit it again, the changes aren't saved on home address page

I can confirm that these happen on all platforms, and are due to the same Form bug: https://github.com/Expensify/App/issues/14784

I used a proposed solution ([this one](https://github.com/Expensify/App/issues/14784#issuecomment-1415302899)) and it fixed the bug

Note: I didn't actually implement the bug fix because the exact fix is being reviewed in that PR - I bumped them to move quicker since the bug is affecting all pages with Form

eVoloshchak commented:
@Beamanator, thank you!
I've re-tested with [this](https://github.com/Expensify/App/issues/14784#issuecomment-1415302899) proposal applied and it works good

Beamanator commented:
Thanks @eVoloshchak ! All you @MonilBhavsar @arosiclair 👍

MonilBhavsar reviewed (commented):
@Beamanator looks good to me and works well
Are you able to reproduce this? I think this is because of large account and sequential requests (or slow internet connection?) cc @eVoloshchak 


https://user-images.githubusercontent.com/32012005/216979054-f6f93b34-865c-49b5-8255-58bcdc004328.mov

Beamanator commented:
@MonilBhavsar I do see a slight delay on the Personal Details Initial Page (when loading the latest private personal details) but definitely not that big of a delay. I wonder if there's a larger delay since this data is encrypted and has to come all the way from Auth? If there's consistently long delays maybe we can have some kind of loading state on that page

Beamanator commented:
I'm merging b/c I got 2 approvals and everything seems like it has been tested pretty thoroughly 👍 Thanks y'all!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

MonilBhavsar commented:
>  If there's consistently long delays maybe we can have some kind of loading state on that page

Yes, a loader sounds good. This might seem broken, but in real we're fetching data

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4116663139) to staging by https://github.com/Beamanator in version: 1.2.67-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

mountiny commented:
Noting there was an issue created by QA when testing this PR https://github.com/Expensify/App/issues/14949 but I will go ahead with the deploy as it seems like its only affecting one platform and this is a big PR so reverting is not a good option. However, it would be great to fix that one sooner than later.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4129309242) to production by https://github.com/mountiny in version: 1.2.67-7 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

aldo-expensify commented on src/components/AddressSearch.js:
```diff
@@ -68,19 +75,27 @@ const defaultProps = {
         zipCode: 'addressZipCode',
     },
+
+    // A maximum of -1 means the user can enter as many characters as they want
+    maxInputLength: -1,
```
bugzero: This default doesn't work well in native and caused [this regression](https://github.com/Expensify/App/issues/14918). We replaced this value by `undefined`.

marcaaron commented on src/libs/actions/PersonalDetails.js:
```diff
@@ -173,6 +173,74 @@ function updateDisplayName(firstName, lastName) {
+                    zip,
+                    country,
+                },
+            },
```
Is there no validation for these address fields? Like if we update it while we are offline and it is a bad address what happens here?

sobitneupane commented on src/pages/settings/Profile/PersonalDetails/AddressPage.js:
```diff
@@ -0,0 +1,226 @@
+        ];
+
+        // Check "State" dropdown is a valid state if selected Country is USA.
+        if (this.state.isUsaForm && !COMMON_CONST.STATES[values.state]) {
```
[This issue](https://github.com/Expensify/App/issues/21793) was not handled in this PR.

Issue: Error is shown in state/province field of Home address, Personal details while changing to a different location than US.

alitoshmatov commented on src/libs/actions/PersonalDetails.js:
```diff
@@ -173,6 +173,74 @@ function updateDisplayName(firstName, lastName) {
+ * @param {String} legalFirstName
+ * @param {String} legalLastName
+ */
+function updateLegalName(legalFirstName, legalLastName) {
+    API.write('UpdateLegalName', {legalFirstName, legalLastName}, {
+        optimisticData: [{
+            onyxMethod: CONST.ONYX.METHOD.MERGE,
+            key: ONYXKEYS.PRIVATE_PERSONAL_DETAILS,
+            value: {
+                legalFirstName,
+                legalLastName,
+            },
+        }],
+    });
+    Navigation.navigate(ROUTES.SETTINGS_PERSONAL_DETAILS);
+}
```
Not sure if this is correct place for this 🤔 but everything lead here. 

Update legal name should also update display name fixed here: https://github.com/Expensify/App/issues/52900
